api/src/main/java/com/github/eyefloaters/console/api/ClientFactory.java

```diff
@@ -210,20 +169,19 @@ Map<String, Object> buildConfiguration(Kafka cluster, ListenerStatus listenerSta
         return config;
     }
 
-    Optional<String> extractCredentials(Optional<String> authorizationHeader) {
-        return authorizationHeader
-                .filter(Objects::nonNull)
-                .filter(authn -> authn.startsWith("Basic "))
-                .map(authn -> authn.substring("Basic ".length()))
-                .map(Base64.getDecoder()::decode)
-                .map(String::new)
-                .filter(authn -> authn.indexOf(':') >= 0)
-                .map(authn -> new String[] {
-                    authn.substring(0, authn.indexOf(':')),
-                    authn.substring(authn.indexOf(':') + 1)
-                })
-                .filter(credentials -> !credentials[0].isEmpty() && !credentials[1].isEmpty())
-                .map(credentials -> String.format(SASL_PLAIN_CONFIG_TEMPLATE, credentials[0], credentials[1]));
-    }
+    void configureOAuthBearer(Map<String, Object> config) {
+        log.debug("SASL/OAUTHBEARER enabled");
+        config.put(SaslConfigs.SASL_MECHANISM, OAUTHBEARER);
+        config.put(SaslConfigs.SASL_LOGIN_CALLBACK_HANDLER_CLASS, "io.strimzi.kafka.oauth.client.JaasClientOauthLoginCallbackHandler");
+        // Do not attempt token refresh ahead of expiration (ExpiringCredentialRefreshingLogin)
+        // May still cause warnings to be logged when token will expired in less than SASL_LOGIN_REFRESH_MIN_PERIOD_SECONDS.
+        config.put(SaslConfigs.SASL_LOGIN_REFRESH_BUFFER_SECONDS, "0");
 
+        final String accessToken = Optional.ofNullable(headers.getHeaderString(HttpHeaders.AUTHORIZATION))
+                .filter(header -> header.regionMatches(true, 0, BEARER, 0, BEARER.length()))
+                .map(header -> header.substring(BEARER.length()))
+                .orElseThrow(() -> new NotAuthorizedException(BEARER.trim()));
```

Comment:
Is "Bearer" enough of an exception message to be useful?

Reply:
This constructor argument is a "challenge" [1] that will ultimately end up in a `WWW-Authenticate` header [2]. In the real world this code will not execute because Quarkus will have already verified that a bearer token exists and is valid per the configured JWKS/Keycloak. This aspect is disabled in the integration tests.

[1] https://jakarta.ee/specifications/restful-ws/3.1/apidocs/jakarta.ws.rs/jakarta/ws/rs/notauthorizedexception#%3Cinit%3E(java.lang.Object,java.lang.Object...)
[2] https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate

Reply:
Thanks, I'm missed that this was a ws exception. That's not something I'm especially familiar with, so thanks for the explanation.